controllers/helloworld_controller.go

```diff
@@ -0,0 +1,201 @@
+
+	if err := r.Get(ctx, svcNamespacedName, svc); err != nil {
+		if errors.IsNotFound(err) {
+			Logging(req, "Creating a service")
+			r.Create(ctx, r.svcForHelloworld(helloworld))
+			return ctrl.Result{Requeue: true}, nil
+		}
+		log.Log.Error(err, "unable to fetch "+helloworld.Name+"'s service")
+		return ctrl.Result{}, err
+	}
+
+	Logging(req, "Service exist - "+svc.Name)
+
+	requiredReplicas := helloworld.Spec.ReplicaCount
+
+	Logging(req, "Required count of pods is "+fmt.Sprint(requiredReplicas))
+
+	podList := &v1.PodList{}
+	opts := []client.ListOption{
+		client.InNamespace(helloworld.Namespace),
+		client.MatchingLabels{"hw_name": helloworld.Name},
+	}
+
+	if err := r.List(ctx, podList, opts...); err != nil {
+		log.Log.Error(err, "unable to list "+helloworld.Name+"'s pods")
+		return ctrl.Result{}, err
+	}
+
+	currentReplicas := len(podList.Items)
+	Logging(req, "Current count of pods is "+fmt.Sprint(currentReplicas))
+
+	if requiredReplicas > int32(currentReplicas) {
+		for i := int32(0); i < requiredReplicas-int32(currentReplicas); i++ {
+			Logging(req, "Creating a pod number "+fmt.Sprint(i))
+			r.Create(ctx, r.podForHelloworld(helloworld))
+		}
+
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	countReady := 0
+	for _, pod := range podList.Items {
+		conditions := pod.Status.Conditions
+		for _, condition := range conditions {
+			if condition.Type == "Ready" && condition.Status == "True" {
+				countReady++
+			}
+		}
+	}
+
+	helloworld.Status.AvailableReplicas = int32(currentReplicas)
+	helloworld.Status.ReadyReplicas = int32(countReady)
+
+	Logging(req, "Updating status")
+
+	if err := r.Status().Update(ctx, helloworld); err != nil {
+		log.Log.Error(err, "unable to update Helloworld")
+		return ctrl.Result{}, err
+	}
+
+	Logging(req, "Finished succesfully")
+
+	return ctrl.Result{Requeue: false}, nil
+}
+
+func (r *HelloworldReconciler) svcForHelloworld(h *v1alpha1.Helloworld) *v1.Service {
+	lbls := labelsForApp(h.Name)
+
+	svc := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      h.Name + "-service",
+			Namespace: h.Namespace,
+			Labels:    lbls,
+		},
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{
+				{
+					Protocol:   v1.ProtocolTCP,
+					Port:       8080,
+					TargetPort: intstr.IntOrString{Type: intstr.Int, IntVal: 8080},
+				}},
+			Selector: lbls,
+			Type:     v1.ServiceTypeClusterIP,
+		},
+	}
+
+	controllerutil.SetControllerReference(h, svc, r.Scheme)
+	return svc
+}
+
+func (r *HelloworldReconciler) podForHelloworld(h *v1alpha1.Helloworld) *v1.Pod {
+	lbls := labelsForApp(h.Name)
+	podName := h.Name + "-pod-" + fmt.Sprint(time.Now().UnixNano())
+
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      podName,
+			Namespace: h.Namespace,
+			Labels:    lbls,
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{{Name: podName + "-contatiner", Image: "gcr.io/run-ai-lab/dans-service", Env: []v1.EnvVar{{Name: "content", Value: h.Spec.DefaultContent}}}},
```

Comment:
the image name should be a const,
or even better - a param for the whole container